File: packages/text-recognition/android/src/main/kotlin/com/margelo/nitro/nitromlkit/text/HybridTextRecognizer.kt
package com.margelo.nitro.nitromlkit.text

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Rect
import android.net.Uri
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.Text
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.tasks.await
import java.net.URL

/**
 * Native Android implementation of TextRecognizer.
 * Extends the Nitrogen-generated HybridTextRecognizerSpec abstract class.
 *
 * Uses MLKit's bundled Latin text-recognition model. Returns the full text plus
 * a structured block -> line -> element hierarchy with bounding boxes.
 */
class HybridTextRecognizer : HybridTextRecognizerSpec() {

  private val recognizer by lazy {
    TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)
  }

  // MARK: - Spec methods

  override fun recognize(imageUri: String): Promise<RecognizedText> {
    return Promise.async { recognizeImage(imageUri) }
  }

  override fun recognizeText(imageUri: String): Promise<String> {
    return Promise.async { recognizeImage(imageUri).text }
  }

  override fun recognizeBatch(imageUris: Array<String>, concurrency: Double): Promise<Array<BatchTextResult>> {
    return Promise.async {
      val maxConcurrent = concurrency.toInt().coerceAtLeast(1)
      val results = mutableListOf<BatchTextResult>()
      var start = 0
      while (start < imageUris.size) {
        val end = minOf(start + maxConcurrent, imageUris.size)
        val chunk = coroutineScope {
          (start until end).map { idx ->
            async {
              try {
                BatchTextResult(idx.toDouble(), recognizeImage(imageUris[idx]), true, null)
              } catch (e: Exception) {
                BatchTextResult(
                  idx.toDouble(),
                  RecognizedText("", emptyArray()),
                  false,
                  e.message ?: "unknown error",
                )
              }
            }
          }.awaitAll()
        }
        results.addAll(chunk)
        start = end
      }
      results.toTypedArray()
    }
  }

  override fun isAvailable(): Boolean = true

  // MARK: - Private helpers

  private suspend fun recognizeImage(uri: String): RecognizedText {
    val bitmap = loadBitmap(uri) ?: throw Error("Failed to load image: $uri")
    val image = InputImage.fromBitmap(bitmap, 0)
    return mapText(recognizer.process(image).await())
  }

  private fun rect(box: Rect?): TextRect = TextRect(
    x = (box?.left ?: 0).toDouble(),
    y = (box?.top ?: 0).toDouble(),
    width = (box?.width() ?: 0).toDouble(),
    height = (box?.height() ?: 0).toDouble(),
  )

  private fun mapText(t: Text): RecognizedText {
    val blocks = t.textBlocks.map { b ->
      val lines = b.lines.map { l ->
        val elements = l.elements.map { e -> TextElement(e.text, rect(e.boundingBox)) }.toTypedArray()
        TextLine(l.text, rect(l.boundingBox), elements)
      }.toTypedArray()
      TextBlock(b.text, rect(b.boundingBox), lines)
    }.toTypedArray()
    return RecognizedText(t.text, blocks)
  }

  private fun loadBitmap(uri: String): Bitmap? {
    return try {
      when {
        uri.startsWith("file://") -> BitmapFactory.decodeFile(uri.removePrefix("file://"))
        uri.startsWith("/") -> BitmapFactory.decodeFile(uri)
        uri.startsWith("content://") -> {
          val context = NitroModules.applicationContext ?: return null
          context.contentResolver.openInputStream(Uri.parse(uri))?.use { BitmapFactory.decodeStream(it) }
        }
        else -> URL(uri).openStream().use { BitmapFactory.decodeStream(it) }
      }
    } catch (e: Exception) {
      null
    }
  }
}

File: packages/text-recognition/ios/HybridTextRecognizer.swift
import Foundation
import UIKit
import NitroModules
import MLKitTextRecognition
import MLKitVision

/**
 * Native iOS implementation of TextRecognizer.
 * Conforms to the Nitrogen-generated HybridTextRecognizerSpec protocol.
 *
 * Uses MLKit's bundled Latin text-recognition model. Returns the full text plus
 * a structured block -> line -> element hierarchy with bounding boxes. Mirrors
 * the Android (Kotlin) implementation.
 */
class HybridTextRecognizer: HybridTextRecognizerSpec {

    // MARK: - HybridObject boilerplate
    var memorySize: Int { MemoryLayout<HybridTextRecognizer>.size }

    private lazy var recognizer: TextRecognizer =
        TextRecognizer.textRecognizer(options: TextRecognizerOptions())

    // MARK: - Spec methods

    func recognize(imageUri: String) throws -> Promise<RecognizedText> {
        return Promise.async {
            return try await self.recognizeImage(imageUri)
        }
    }

    func recognizeText(imageUri: String) throws -> Promise<String> {
        return Promise.async {
            return try await self.recognizeImage(imageUri).text
        }
    }

    func recognizeBatch(imageUris: [String], concurrency: Double) throws -> Promise<[BatchTextResult]> {
        return Promise.async {
            let maxConcurrent = max(1, Int(concurrency))
            var results = [BatchTextResult]()
            results.reserveCapacity(imageUris.count)

            for chunkStart in stride(from: 0, to: imageUris.count, by: maxConcurrent) {
                let chunkEnd = min(chunkStart + maxConcurrent, imageUris.count)
                let chunk = Array(imageUris[chunkStart..<chunkEnd])

                let chunkResults = await withTaskGroup(of: (Int, BatchTextResult).self) { group in
                    for (i, uri) in chunk.enumerated() {
                        let globalIdx = chunkStart + i
                        group.addTask {
                            do {
                                let text = try await self.recognizeImage(uri)
                                return (globalIdx, BatchTextResult(index: Double(globalIdx), text: text, success: true, error: nil))
                            } catch {
                                return (globalIdx, BatchTextResult(
                                    index: Double(globalIdx),
                                    text: RecognizedText(text: "", blocks: []),
                                    success: false,
                                    error: error.localizedDescription
                                ))
                            }
                        }
                    }
                    var out = [(Int, BatchTextResult)]()
                    for await r in group { out.append(r) }
                    return out.sorted(by: { $0.0 < $1.0 }).map(\.1)
                }
                results.append(contentsOf: chunkResults)
            }
            return results
        }
    }

    func isAvailable() throws -> Bool {
        return true
    }

    // MARK: - Private helpers

    private func recognizeImage(_ uri: String) async throws -> RecognizedText {
        guard let image = loadImage(from: uri) else {
            throw RuntimeError.error(withMessage: "Failed to load image: \(uri)")
        }
        let visionImage = VisionImage(image: image)
        visionImage.orientation = image.imageOrientation
        let result = try await recognizer.process(visionImage)
        return mapText(result)
    }

    private func rect(_ frame: CGRect) -> TextRect {
        return TextRect(
            x: Double(frame.origin.x),
            y: Double(frame.origin.y),
            width: Double(frame.width),
            height: Double(frame.height)
        )
    }

    private func mapText(_ t: MLKitTextRecognition.Text) -> RecognizedText {
        let blocks: [TextBlock] = t.blocks.map { block in
            let lines: [TextLine] = block.lines.map { line in
                let elements: [TextElement] = line.elements.map { element in
                    TextElement(text: element.text, bounds: self.rect(element.frame))
                }
                return TextLine(text: line.text, bounds: self.rect(line.frame), elements: elements)
            }
            return TextBlock(text: block.text, bounds: self.rect(block.frame), lines: lines)
        }
        return RecognizedText(text: t.text, blocks: blocks)
    }

    private func loadImage(from uri: String) -> UIImage? {
        let path: String
        if uri.hasPrefix("file://") {
            path = String(uri.dropFirst(7))
        } else if uri.hasPrefix("/") {
            path = uri
        } else {
            guard let url = URL(string: uri), let data = try? Data(contentsOf: url) else { return nil }
            return UIImage(data: data)
        }
        return UIImage(contentsOfFile: path)
    }
}
